python/pyspark/pandas/frame.py

```diff
@@ -8828,22 +8842,117 @@ def describe(self, percentiles: Optional[List[float]] = None) -> "DataFrame":
-        internal = InternalFrame(
-            spark_frame=sdf,
-            index_spark_columns=[scol_for(sdf, "summary")],
-            column_labels=column_labels,
-            data_spark_columns=[
-                scol_for(sdf, self._internal.spark_column_name_for(label))
-                for label in column_labels
-            ],
-        )
-        return DataFrame(internal).astype("float64")
+            # Get `count` & `unique` for each columns
+            has_timestamp_type = True in is_timestamp_types
+            if not has_timestamp_type:
+                counts, uniques = map(
+                    lambda x: x[1:], sdf.summary("count", "count_distinct").take(2)
+                )
+            else:
+                # `summary` doesn't support for timestamp column, so we should manually compute it
+                # if timestamp type column exists.
+                counts = []
+                uniques = []
+                exprs = []
+                for column in exprs_non_numeric:
+                    exprs.append(F.count(column))
+                    exprs.append(F.count_distinct(column))
+
+                count_unique_values = list(sdf.select(*exprs).first().asDict().values())
+                for i in range(0, len(count_unique_values) - 1, 2):
+                    counts.append(str(count_unique_values[i]))
+                    uniques.append(str(count_unique_values[i + 1]))
+
+            # Get `top` & `freq` for each columns
+            tops = []
+            freqs = []
+            for column in exprs_non_numeric:
+                top, freq = (
+                    sdf.groupby(column)
+                    .count()
+                    .sort("count", ascending=False)
+                    .first()
+                    .asDict()
+                    .values()
+                )
+                tops.append(str(top))
+                freqs.append(str(freq))
+
+            stats = [counts, uniques, tops, freqs]
+            stats_names = ["count", "unique", "top", "freq"]
+
+            # Get `first` & `last` for each columns if timestamp type column exists.
+            if has_timestamp_type:
+                exprs = []
+                for is_timestamp_type, column, column_name in zip(
+                    is_timestamp_types, exprs_non_numeric, column_names
+                ):
+                    if is_timestamp_type:
+                        # `first` & `last` are min & max respectively for timestamp type.
+                        exprs.append(F.min(column))
+                        exprs.append(F.max(column))
+                    else:
+                        # `first` & `last` are always NaN for string type columns.
+                        exprs.append(F.lit(np.NaN).alias(column_name + "_first"))
+                        exprs.append(F.lit(np.NaN).alias(column_name + "_last"))
+
+                firsts = []
+                lasts = []
+                first_last_values = list(sdf.select(*exprs).first().asDict().values())
```

Comment:
I think we don't need to convert it to `asDict`. just call list(sdf.select(*exprs).first()) should be fine. How does it work when `sdf` is empty?

Reply:
Thanks! I'll address it

I believe `sdf` cannot be empty since it raises `ValueError("Cannot describe a DataFrame without columns")` when the `exprs_non_numeric` is empty (`sdf` is created by selecting `exprs_non_numeric`)

```python
        if len(exprs_numeric) == 0:
            if len(exprs_non_numeric) == 0:
                raise ValueError("Cannot describe a DataFrame without columns")
```

Reply:
Thanks! I'll address it

I believe `sdf` cannot be empty since it raises `ValueError("Cannot describe a DataFrame without columns")` before this place, when the `exprs_non_numeric` is empty (`sdf` is created by selecting `exprs_non_numeric`)

```python
        if len(exprs_numeric) == 0:
            if len(exprs_non_numeric) == 0:
                raise ValueError("Cannot describe a DataFrame without columns")
```

Reply:
what about when rows are empty?

Reply:
Ah, do you think about the case when the `sdf.select(*exprs)` is empty ??

If so, I believe this also cannot be empty because `exprs` is the list of `max` & `min` value for each columns in `sdf`. (`sdf` is always not empty, thus the `min` & `max` value for `sdf` is always existing)

Reply:
👌

Reply:
btw, can we avoid using `asDict().values()`? this is unnecessary. you can just get it from `list(row)`

Reply:
Sure, let me address the `asDict` usage here and there. Thanks!